File: src/main/java/commands/listeners/OnMessageInputListener.java
package commands.listeners;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Function;
import commands.Command;
import commands.CommandContainer;
import commands.CommandListenerMeta;
import core.MainLogger;
import core.MemberCacheController;
import core.utils.BotPermissionUtil;
import core.utils.ExceptionUtil;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;

public interface OnMessageInputListener {

    MessageInputResponse onMessageInput(GuildMessageReceivedEvent event, String input) throws Throwable;

    EmbedBuilder draw(Member member) throws Throwable;

    default void registerMessageInputListener(Member member) {
        registerMessageInputListener(member, true);
    }

    default void registerMessageInputListener(Member member, boolean draw) {
        Command command = (Command) this;
        registerMessageInputListener(member, draw, event -> {
                    boolean ok = event.getMember().getIdLong() == member.getIdLong() &&
                            event.getChannel().getIdLong() == command.getTextChannelId().orElse(0L);
                    return ok ? CommandListenerMeta.CheckResponse.ACCEPT : CommandListenerMeta.CheckResponse.IGNORE;
                }
        );
    }

    default void registerMessageInputListener(Member member, boolean draw, Function<GuildMessageReceivedEvent, CommandListenerMeta.CheckResponse> validityChecker) {
        Command command = (Command) this;

        Runnable onTimeOut = () -> {
            try {
                command.deregisterListeners();
                command.onListenerTimeOutSuper();
            } catch (Throwable throwable) {
                MainLogger.get().error("Exception on time out", throwable);
            }
        };

        Runnable onOverridden = () -> {
            try {
                onMessageInputOverridden();
            } catch (Throwable throwable) {
                MainLogger.get().error("Exception on overridden", throwable);
            }
        };

        CommandListenerMeta<GuildMessageReceivedEvent> commandListenerMeta =
                new CommandListenerMeta<>(member.getIdLong(), validityChecker, onTimeOut, onOverridden, command);
        CommandContainer.registerListener(OnMessageInputListener.class, commandListenerMeta);

        try {
            if (draw && command.getDrawMessageId().isEmpty()) {
                EmbedBuilder eb = draw(member);
                if (eb != null) {
                    command.drawMessage(eb);
                }
            }
        } catch (Throwable e) {
            command.getTextChannel().ifPresent(channel -> {
                ExceptionUtil.handleCommandException(e, command, channel);
            });
        }
    }

    default MessageInputResponse processMessageInput(GuildMessageReceivedEvent event) {
        Command command = (Command) this;
        AtomicBoolean isProcessing = new AtomicBoolean(true);

        command.addLoadingReaction(event.getMessage(), isProcessing);
        try {
            if (command.getCommandProperties().requiresFullMemberCache()) {
                MemberCacheController.getInstance().loadMembersFull(event.getGuild()).get();
            }
            MessageInputResponse messageInputResponse = onMessageInput(event, event.getMessage().getContentRaw());
            if (messageInputResponse != null) {
                if (messageInputResponse == MessageInputResponse.SUCCESS) {
                    CommandContainer.refreshListeners(command);
                    if (BotPermissionUtil.can(event.getChannel(), Permission.MESSAGE_MANAGE)) {
                        event.getMessage().delete().queue();
                    }
                }

                EmbedBuilder eb = draw(event.getMember());
                if (eb != null) {
                    ((Command) this).drawMessage(eb);
                }
            }
            return messageInputResponse;
        } catch (Throwable e) {
            ExceptionUtil.handleCommandException(e, command, event.getChannel());
            return MessageInputResponse.ERROR;
        } finally {
            isProcessing.set(false);
        }
    }

    default void onMessageInputOverridden() throws Throwable {
    }

}

File: src/main/java/commands/listeners/OnMessageInputListener.kt
package commands.listeners

import commands.Command
import commands.CommandContainer
import commands.CommandListenerMeta
import commands.CommandListenerMeta.CheckResponse
import core.ExceptionLogger
import core.MainLogger
import core.MemberCacheController
import core.utils.BotPermissionUtil
import core.utils.ExceptionUtil
import mysql.hibernate.entity.guild.GuildEntity
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import java.util.concurrent.atomic.AtomicBoolean

interface OnMessageInputListener : Drawable {

    @Throws(Throwable::class)
    fun onMessageInput(event: MessageReceivedEvent, input: String): MessageInputResponse?

    fun registerMessageInputListener(member: Member, draw: Boolean = true) {
        val command = this as Command
        registerMessageInputListener(member, draw) { event ->
            val ok = event.member!!.idLong == member.idLong &&
                    event.channel.idLong == command.textChannelId.orElse(0L)
            if (ok) CheckResponse.ACCEPT else CheckResponse.IGNORE
        }
    }

    fun registerMessageInputListener(member: Member, draw: Boolean, validityChecker: (MessageReceivedEvent) -> CheckResponse) {
        val command = this as Command
        val onTimeOut = {
            try {
                command.refreshGuildEntity().use {
                    command.deregisterListeners()
                    command.onListenerTimeOutSuper()
                }
            } catch (throwable: Throwable) {
                MainLogger.get().error("Exception on time out", throwable)
            }
        }
        val onOverridden = {
            try {
                command.refreshGuildEntity().use {
                    onMessageInputOverridden()
                }
            } catch (throwable: Throwable) {
                MainLogger.get().error("Exception on overridden", throwable)
            }
        }
        val commandListenerMeta = CommandListenerMeta(member.idLong, validityChecker, onTimeOut, onOverridden, command)
        CommandContainer.registerListener(OnMessageInputListener::class.java, commandListenerMeta)
        try {
            if (draw && command.drawMessageId.isEmpty) {
                val eb = draw(member)
                if (eb != null) {
                    command.drawMessage(eb)
                        .exceptionally(ExceptionLogger.get())
                }
            }
        } catch (e: Throwable) {
            command.textChannel.ifPresent { ExceptionUtil.handleCommandException(e, command, commandEvent, guildEntity) }
        }
    }

    fun processMessageInput(event: MessageReceivedEvent, guildEntity: GuildEntity): MessageInputResponse? {
        val command = this as Command
        val isProcessing = AtomicBoolean(true)
        try {
            if (command.commandProperties.requiresFullMemberCache) {
                MemberCacheController.getInstance().loadMembersFull(event.guild).get()
            }
            command.guildEntity = guildEntity
            val messageInputResponse = onMessageInput(event, event.message.contentRaw)
            if (messageInputResponse != null) {
                if (messageInputResponse === MessageInputResponse.SUCCESS) {
                    CommandContainer.refreshListeners(command)
                    if (BotPermissionUtil.can(event.guildChannel, Permission.MESSAGE_MANAGE)) {
                        event.message.delete().queue()
                    }
                }
                val eb = draw(event.member!!)
                if (eb != null) {
                    (this as Command).drawMessage(eb).exceptionally(ExceptionLogger.get())
                }
            }
            return messageInputResponse
        } catch (e: Throwable) {
            ExceptionUtil.handleCommandException(e, command, commandEvent, guildEntity)
            return MessageInputResponse.ERROR
        } finally {
            isProcessing.set(false)
        }
    }

    @Throws(Throwable::class)
    fun onMessageInputOverridden() {
    }

}